File: typescript/tests/storage-memory/find-by-licensable.test.ts
/**
 * findLicensesByLicensable — memory adapter coverage.
 *
 * Same scenarios as findByLicensable in the spec: matches across scopes,
 * scope-filtered lookup, ordering, no-match.
 */

import { describe, expect, it } from 'bun:test';

import { MemoryStorage } from '../../src/storage/memory/index.ts';
import type { LicenseInput } from '../../src/storage/types.ts';
import type { LicenseStatus, UUIDv7 } from '../../src/types.ts';

let counter = 0;
function freshKey(): string {
  counter++;
  return `LIC-AAAA-BBBB-CCCC-${counter.toString().padStart(4, '0')}`;
}

function inputFor(
  type: string,
  id: string,
  opts: Partial<LicenseInput> & { status?: LicenseStatus } = {},
): LicenseInput {
  return {
    scope_id: opts.scope_id ?? null,
    template_id: null,
    licensable_type: type,
    licensable_id: id,
    license_key: freshKey(),
    status: opts.status ?? 'active',
    max_usages: 5,
    activated_at: null,
    expires_at: null,
    grace_until: null,
    meta: {},
    ...opts,
  };
}

describe('memory adapter — findLicensesByLicensable', () => {
  it('returns every license attached to a (type, id) across scopes', async () => {
    const s = new MemoryStorage();
    const scope1 = await s.createScope({ slug: 's1', name: 'Scope 1', meta: {} });
    const scope2 = await s.createScope({ slug: 's2', name: 'Scope 2', meta: {} });
    await s.createLicense(inputFor('User', 'u_1', { scope_id: scope1.id }));
    await s.createLicense(inputFor('User', 'u_1', { scope_id: scope2.id }));
    await s.createLicense(inputFor('User', 'u_1')); // global scope
    await s.createLicense(inputFor('User', 'u_2', { scope_id: scope1.id }));

    const matches = await s.findLicensesByLicensable({ type: 'User', id: 'u_1' });
    expect(matches.length).toBe(3);
  });

  it('scope_id filter narrows the result', async () => {
    const s = new MemoryStorage();
    const scope1 = await s.createScope({ slug: 's1', name: 'Scope 1', meta: {} });
    await s.createLicense(inputFor('User', 'u_1', { scope_id: scope1.id }));
    await s.createLicense(inputFor('User', 'u_1'));

    const scoped = await s.findLicensesByLicensable({
      type: 'User',
      id: 'u_1',
      scope_id: scope1.id,
    });
    expect(scoped.length).toBe(1);
    expect(scoped[0]?.scope_id).toBe(scope1.id);
  });

  it('scope_id: null filters to global-scope licenses only', async () => {
    const s = new MemoryStorage();
    const scope1 = await s.createScope({ slug: 's1', name: 'Scope 1', meta: {} });
    await s.createLicense(inputFor('User', 'u_1', { scope_id: scope1.id }));
    await s.createLicense(inputFor('User', 'u_1'));

    const globalOnly = await s.findLicensesByLicensable({
      type: 'User',
      id: 'u_1',
      scope_id: null,
    });
    expect(globalOnly.length).toBe(1);
    expect(globalOnly[0]?.scope_id).toBeNull();
  });

  it('returns empty array for unknown licensable', async () => {
    const s = new MemoryStorage();
    const result = await s.findLicensesByLicensable({ type: 'User', id: 'u_unknown' });
    expect(result).toEqual([]);
  });

  it('sorts by created_at DESC', async () => {
    const s = new MemoryStorage();
    const a = await s.createLicense(inputFor('User', 'u_x'));
    // Force a tick so timestamps differ.
    await new Promise((r) => setTimeout(r, 5));
    const b = await s.createLicense(
      inputFor('User', 'u_x', {
        scope_id: (await s.createScope({ slug: 'sx', name: 'X', meta: {} })).id as UUIDv7,
      }),
    );
    const matches = await s.findLicensesByLicensable({ type: 'User', id: 'u_x' });
    expect(matches.map((m) => m.id)).toEqual([b.id, a.id]);
  });
});

File: licensing/storage/memory/find_by_licensable_test.go
package memory_test

import (
	"testing"

	lic "github.com/AnoRebel/licensing/licensing"
	"github.com/AnoRebel/licensing/licensing/storage/memory"
)

// Mirrors typescript/tests/storage-memory/find-by-licensable.test.ts.

func licInput(t lic.LicenseStatus, scopeID *string, lt, lid, key string) lic.LicenseInput {
	return lic.LicenseInput{
		ScopeID:        scopeID,
		LicensableType: lt,
		LicensableID:   lid,
		LicenseKey:     key,
		Status:         t,
		MaxUsages:      5,
		Meta:           map[string]any{},
	}
}

func TestMemory_FindByLicensable_AcrossScopes(t *testing.T) {
	s := memory.New(memory.Options{})
	scope1, _ := s.CreateScope(lic.LicenseScopeInput{Slug: "s1", Name: "Scope 1", Meta: map[string]any{}})
	scope2, _ := s.CreateScope(lic.LicenseScopeInput{Slug: "s2", Name: "Scope 2", Meta: map[string]any{}})

	_, _ = s.CreateLicense(licInput(lic.LicenseStatusActive, &scope1.ID, "User", "u1", "LIC-A"))
	_, _ = s.CreateLicense(licInput(lic.LicenseStatusActive, &scope2.ID, "User", "u1", "LIC-B"))
	_, _ = s.CreateLicense(licInput(lic.LicenseStatusActive, nil, "User", "u1", "LIC-C")) // global
	_, _ = s.CreateLicense(licInput(lic.LicenseStatusActive, &scope1.ID, "User", "u2", "LIC-D"))

	matches, err := s.FindLicensesByLicensable(lic.FindByLicensableQuery{
		Type: "User",
		ID:   "u1",
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(matches) != 3 {
		t.Fatalf("expected 3 matches, got %d", len(matches))
	}
}

func TestMemory_FindByLicensable_ScopedFilter(t *testing.T) {
	s := memory.New(memory.Options{})
	scope, _ := s.CreateScope(lic.LicenseScopeInput{Slug: "s1", Name: "S1", Meta: map[string]any{}})
	_, _ = s.CreateLicense(licInput(lic.LicenseStatusActive, &scope.ID, "User", "u1", "LIC-A"))
	_, _ = s.CreateLicense(licInput(lic.LicenseStatusActive, nil, "User", "u1", "LIC-B"))

	scoped, err := s.FindLicensesByLicensable(lic.FindByLicensableQuery{
		Type: "User", ID: "u1", ScopeID: &scope.ID, ScopeIDSet: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(scoped) != 1 {
		t.Fatalf("expected 1 scoped match, got %d", len(scoped))
	}
	if scoped[0].ScopeID == nil || *scoped[0].ScopeID != scope.ID {
		t.Fatalf("expected scoped match, got %v", scoped[0].ScopeID)
	}

	globalOnly, err := s.FindLicensesByLicensable(lic.FindByLicensableQuery{
		Type: "User", ID: "u1", ScopeID: nil, ScopeIDSet: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(globalOnly) != 1 {
		t.Fatalf("expected 1 global match, got %d", len(globalOnly))
	}
	if globalOnly[0].ScopeID != nil {
		t.Fatalf("expected nil scope on global match, got %v", globalOnly[0].ScopeID)
	}
}

func TestMemory_FindByLicensable_NotFound(t *testing.T) {
	s := memory.New(memory.Options{})
	matches, err := s.FindLicensesByLicensable(lic.FindByLicensableQuery{
		Type: "User", ID: "u_nope",
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(matches) != 0 {
		t.Fatalf("expected 0 matches, got %d", len(matches))
	}
}
